### PR TITLE
enh(behat): replace selenium by panther

### PR DESCRIPTION
**Pull request by kduret**
## Description

replace selenium by panther

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x
- [x] 23.04.x (master)

### CODE REVIEW

**Review by jeremyjaouen — APPROVED** [review]

File: src/behat/CentreonContext.php [hunk 1]
```diff
@@ -57,6 +60,72 @@ public function __construct($parameters = array())
...
+            $kernelOptions = []; # unused cause we do not extend class KernelTestCase
+
+            $managerOptions = [
+                'goog:chromeOptions' => $chromeArgs,
+            ];
+
+            $driver = new PantherDriver($defaultOptions, $kernelOptions, $managerOptions);
+            $driver->start();
+        } catch (\Exception $e) {
+            throw new \Exception("Cannot instantiate mink driver.\n" . $e->getMessage());
```
Comment (callapa) [inline src/behat/CentreonContext.php:114]: mink or Panther driver ?

Comment (kduret) [inline src/behat/CentreonContext.php:114]:
```suggestion
            throw new \Exception("Cannot instantiate panther driver.\n" . $e->getMessage());
```

**Review by callapa — APPROVED** [review]

**Review by abuathier — APPROVED** [review]

**Review by TamazC — APPROVED** [review]

**Comment by kduret** [comment]
waiting to migrate BAM and PPM tests before merging this PR

File: src/behat/Configuration/PollerConfigurationExportPage.php [hunk 2]
```diff
@@ -113,14 +113,25 @@ public function setPollers($pollers)
...
                     function () {
-                        return !$this->context->assertFind('css', '.centreon-popin')->isVisible();
-                    },
-                    'Select all confirmation popin did not close properly.'
+                        return count(
+                            $this->context->getSession()->getPage()->findAll(
+                                'css',
+                                '.select2-container--open li.select2-results__option'
+                            )
+                        ) == 0;
```
Comment (kduret) [inline src/behat/Configuration/PollerConfigurationExportPage.php:133]:
```suggestion
                        return empty(
                            $this->context->getSession()->getPage()->findAll(
                                'css',
                                '.select2-container--open li.select2-results__option'
                            )
                        );
```

File: src/behat/SpinTrait.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Waiting an action
+     *
+     * @param closure $closure The function to execute for test the loading.
+     * @param string $timeoutMsg The custom message on timeout.
+     * @param int $wait The timeout in seconds.
+     * @return bool
+     * @throws \Exception
+     */
+    public function spin($closure, $timeoutMsg = 'Load timeout', $wait = 60)
```
Comment (abuathier) [inline src/behat/SpinTrait.php:34]: TypeHinting ?
```suggestion
     * @param callable $closure The function to execute for test the loading.
     * @param string $timeoutMsg The custom message on timeout.
     * @param int $wait The timeout in seconds.
     * @return bool
     * @throws \Exception
     */
    public function spin(callable $closure, string $timeoutMsg = 'Load timeout', int $wait = 60)
```

File: src/behat/CentreonContext.php [hunk 4]
```diff
@@ -57,6 +61,79 @@ public function __construct($parameters = array())
         parent::__construct($parameters);
     }
 
+    /**
+     *  Properly set WebDriver driver.
+     */
+    public function setContainerWebDriver()
```
Comment (abuathier) [inline src/behat/CentreonContext.php:67]:
```suggestion
    public function setContainerWebDriver(): void
```

File: src/behat/Container.php [hunk 5]
```diff
@@ -115,8 +127,19 @@ private function exec($cmd)
...
+            $this->spin(
+                function ($context) {
+                    $context->exec('docker-compose -f ' . $this->composeFile . ' -p ' . $this->id . ' kill');
+                    $context->exec('docker-compose -f ' . $this->composeFile . ' -p ' . $this->id . ' down -v');
+                    return true;
+                },
+                'Cannot stop docker containers',
+                30
+            );
+        } catch (\Exception $e) {
```
Comment (kduret) [inline src/behat/Container.php:140]:
```suggestion
        } catch (\Throwable $e) {
```

File: src/behat/SpinTrait.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        $lastException = null;
+        while (time() <= $limit) {
+            try {
+                if ($closure($this)) {
+                    return true;
+                }
+            } catch (SpinStopException $e) {
+                // stop spining
+                throw $e;
+            } catch (\Exception $e) {
```
Comment (kduret) [inline src/behat/SpinTrait.php:46]:
```suggestion
            } catch (\Throwable $e) {
```

File: src/behat/UtilsContext.php [hunk 7]
```diff
@@ -519,14 +453,14 @@ function ($context) {
...
                             return true;
                         }
                     }
                     return false;
                 },
-                'this error will not be displayed'
+                'this error will not be displayed',
+                10
             );
         } catch (\Exception $e) {
```
Comment (kduret) [inline src/behat/UtilsContext.php:465]:
```suggestion
        } catch (\Throwable $e) {
```

File: src/behat/CentreonContext.php [hunk 8]
```diff
@@ -57,6 +61,79 @@ public function __construct($parameters = array())
...
+                ],
+            ];
+
+            $_SERVER['PANTHER_NO_SANDBOX'] = 1;
+            $_SERVER['PANTHER_CHROME_ARGUMENTS'] = implode(' ', $chromeArgs);
+
+            $driver = new PantherDriver($defaultOptions, $kernelOptions, $managerOptions);
+            $driver->start();
+        } catch (\Exception $e) {
+            throw new \Exception("Cannot instantiate panther driver : " . $e->getMessage());
```
Comment (kduret) [inline src/behat/CentreonContext.php:122]:
```suggestion
            throw new \Exception("Cannot instantiate panther driver : " . $e->getMessage(), (int) $e->getCode(), $e);
```

File: src/behat/CentreonContext.php [hunk 9]
```diff
@@ -57,6 +61,79 @@ public function __construct($parameters = array())
...
+        }
+
+        try {
+            $session = new Session($driver);
+            $mink = new Mink([
+                'panther' => $session,
+            ]);
+            $mink->setDefaultSessionName('panther');
+            $this->setMink($mink);
+        } catch (\Exception $e) {
```
Comment (kduret) [inline src/behat/CentreonContext.php:132]:
```suggestion
        } catch (\Throwable $e) {
```

File: src/behat/CentreonContext.php [hunk 10]
```diff
@@ -57,6 +61,79 @@ public function __construct($parameters = array())
...
+            throw new \Exception("Cannot register mink session.\n" . $e->getMessage());
```
Comment (kduret) [inline src/behat/CentreonContext.php:133]:
```suggestion
            throw new \Exception("Cannot register mink session.\n" . $e->getMessage(), (int) $e->getCode(), $e);
```

File: src/behat/ConfigurationPage.php [hunk 11]
```diff
@@ -152,10 +152,12 @@ public function setProperties($properties)
                     $this->context->selectInList($propertyLocator, $value);
                     break;
                 case 'select2':
-                    $value = is_array($value) ? $value : array($value);
                     $this->context->emptySelectTwo($propertyLocator);
-                    foreach ($value as $element) {
-                        $this->context->selectToSelectTwo($propertyLocator, $element);
+                    if (!empty($value)) {
+                        $value = is_array($value) ? $value : array($value);
```
Comment (abuathier) [inline src/behat/ConfigurationPage.php:157]:
```suggestion
                        $value = (array) $value;
```

**Review by dmyios — APPROVED** [review]

**Pull request merged**